Public/Src/Cache/ContentStore/Library/Stores/ContentHashAddressableSnapshot.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using BuildXL.Cache.ContentStore.Hashing;
+
+namespace BuildXL.Cache.ContentStore.Stores
+{
+    /// <summary>
+    /// This class is specifically for usage within <see cref="ContentHashAddressableSnapshot{T}"/>. It is a reference type on purpose, in order to avoid issues
+    /// with the maximum object size limit.
+    /// </summary>
+    /// <typeparam name="T">Type tagged with a hash</typeparam>
+    public class PayloadFromDisk<T> : IComparable<PayloadFromDisk<T>>

```

Comment:
BTW, I would slightly prefer to have an external comparer instead of making the payload comparable.

When a class is comparable, then one may think that it should be equatable as well. But this is quite a bit of unncessary work. So I would suggest to use a separate comparer that will compare this instance by Hash.

Reply:
This was done on purpose. It is more error-prone to duplicate code, and code won't compile if you attempt to compare, so it won't go silently.

---
In reply to: [285317541](https://github.com/microsoft/BuildXL/pull/329#discussion_r285317541) [](ancestors = 285317541)

Reply:
I'm fine with this approach, but the alternative is not more error-prone:

public class ByHashPayloadFromDiskComparer<T> : IComparer<PayloadFromDisk<T>> {
  public int Compare(PayloadFromDisk<T> left, PayloadFromDisk<T> right) => left.Hash.CompareTo(right.Hash);
}

---
In reply to: [285320584](https://github.com/microsoft/BuildXL/pull/329#discussion_r285320584) [](ancestors = 285320584,285317541)